File: ios/Sources/CleverTapNativeDisplay/Bridge/ReflectionSeeder.swift
//
//  ReflectionSeeder.swift
//  CleverTapNativeDisplay
//

// MARK: - Older Core SDK Fallback
// Used when -[CleverTap setDisplayUnitCache:] is not available on the host's
// Core SDK version. Reaches the existing CTDisplayUnitController via KVC and
// merges the supplied units into its `displayUnits` array so that the
// mandatory cache lookup inside `recordDisplayUnit*EventForID:` succeeds.
//
// Failures are logged once per process and propagated as `false` so callers
// can fall through to a callback-only path.

import Foundation
import ObjectiveC

enum ReflectionSeeder {

    private static var loggedFailure = false

    /// Merges the supplied units into Core SDK's display-unit cache.
    ///
    /// - Parameters:
    ///   - cleverTapInstance: the live CleverTap instance.
    ///   - unitDicts: dictionary representations of `CleverTapDisplayUnit`s
    ///     (typically from `NativeDisplayUnit.rawJson`).
    /// - Returns: `true` if at least one unit was added; `false` on failure.
    @discardableResult
    static func seed(cleverTapInstance: NSObject, unitDicts: [[String: Any]]) -> Bool {
        guard !unitDicts.isEmpty else { return false }
        guard let controller = cleverTapInstance.value(forKey: "displayUnitController") as? NSObject else {
            return logOnce("displayUnitController not reachable via KVC")
        }
        guard NSClassFromString("CleverTapDisplayUnit") is NSObject.Type else {
            return logOnce("CleverTapDisplayUnit class not available")
        }

        let unitIDKey = "unitID"

        var byID: [String: NSObject] = [:]
        if let existing = controller.value(forKey: "displayUnits") as? [NSObject] {
            for unit in existing {
                if let id = unit.value(forKey: unitIDKey) as? String {
                    byID[id] = unit
                }
            }
        }

        for dict in unitDicts {
            guard let unit = NativeDisplayUnitCacheImpl.instantiateCleverTapDisplayUnit(with: dict) else { continue }
            guard let id = unit.value(forKey: unitIDKey) as? String else { continue }
            byID[id] = unit
        }

        controller.setValue(Array(byID.values), forKey: "displayUnits")
        return true
    }

    @discardableResult
    private static func logOnce(_ message: String) -> Bool {
        guard !loggedFailure else { return false }
        loggedFailure = true
        print("[ReflectionSeeder] \(message); display unit events may be unattributed.")
        return false
    }
}

File: android/sdk/src/main/kotlin/com/clevertap/android/nativedisplay/bridge/ReflectionSeeder.kt
package com.clevertap.android.nativedisplay.bridge

import com.clevertap.android.nativedisplay.internal.NDLogger
import org.json.JSONObject

/**
 * Older-Core-SDK fallback used when [CleverTapAPI.setDisplayUnitCache] is not
 * available on the host's Core SDK version. Mutates the existing
 * `CTDisplayUnitController.items` HashMap directly so that the unit becomes
 * visible to `pushDisplayUnit*EventForID` lookups, without invoking
 * `updateDisplayUnits` (which is replace-mode and would clobber Core SDK's
 * own server-driven entries).
 *
 * All access is reflective; failures are logged once per process and
 * propagated as a `false` return so callers can fall through to a
 * callback-only path.
 */
internal object ReflectionSeeder {

    private const val TAG = "NDReflectionSeeder"

    @Volatile
    private var loggedFailure = false

    /**
     * Seeds the supplied units into the Core SDK's display-unit cache.
     *
     * @return true if at least one unit was added; false on any failure
     * (logged once per process).
     */
    fun seed(cleverTapApi: Any, units: List<JSONObject>): Boolean {
        if (units.isEmpty()) return false
        return try {
            val controller = resolveController(cleverTapApi) ?: return logOnce("controller unreachable")
            @Suppress("UNCHECKED_CAST")
            val items = field(controller, "items") as MutableMap<String, Any>
            val toDisplayUnit = Class.forName("com.clevertap.android.sdk.displayunits.model.CleverTapDisplayUnit")
                .getMethod("toDisplayUnit", JSONObject::class.java)
            val getUnitID = Class.forName("com.clevertap.android.sdk.displayunits.model.CleverTapDisplayUnit")
                .getMethod("getUnitID")
            synchronized(controller) {
                for (json in units) {
                    val unit = toDisplayUnit.invoke(null, json) ?: continue
                    val id = getUnitID.invoke(unit) as? String ?: continue
                    items[id] = unit
                }
            }
            true
        } catch (t: Throwable) {
            logOnce("seed failed: ${t.message}")
        }
    }

    private fun resolveController(cleverTapApi: Any): Any? {
        val coreState = field(cleverTapApi, "coreState")
        val analyticsManager = coreState.javaClass.getMethod("getAnalyticsManager").invoke(coreState) ?: return null
        val controllerManager = field(analyticsManager, "controllerManager")
        // Newer Core SDK exposes getDisplayUnitCache(); older only getCTDisplayUnitController().
        return try {
            controllerManager.javaClass.getMethod("getDisplayUnitCache").invoke(controllerManager)
        } catch (_: NoSuchMethodException) {
            controllerManager.javaClass.getMethod("getCTDisplayUnitController").invoke(controllerManager)
        }
    }

    private fun field(obj: Any, name: String): Any =
        obj.javaClass.getDeclaredField(name).apply { isAccessible = true }.get(obj)
            ?: throw IllegalStateException("Field '$name' on ${obj.javaClass.simpleName} was null")

    // Always returns `false` so callers can `return logOnce(...)` as a "log once
    // and signal failure" shortcut from a Boolean-returning function.
    @Suppress("FunctionOnlyReturningConstant")
    private fun logOnce(message: String): Boolean {
        if (!loggedFailure) {
            loggedFailure = true
            NDLogger.w(TAG, "$message; events will be unattributed.")
        }
        return false
    }
}
